projects/file-parsing/main.go

```diff
@@ -0,0 +1,360 @@
+	}
+
+	// if the data slice is empty
+	if len(*dataSlice) == 0 {
+		return errors.New("[3] could not parse any data in CSV format")
+	}
+
+	return nil
+}
+
+// [4] BINARY
+func parseBinary(file []byte, dataSlice *[]Player) error {
+	// each record contains exactly four bytes
+	// representing the score as a signed 32-bit integer (in the above endian format)
+	// then the name of the player stored in UTF-8
+	// which may not contain a null character, followed by a null terminating character
+
+	// --- DEBUG PRINT TABLE
+	// for index := 0; index < len(file); index++ {
+	// 	// print out the bytes in the slice in various formats
+	// 	fmt.Printf("slice index: %2d | Binary: %08b | Bytes (as an Integer): %3v | Hexadecimal: %2x | Decimal: %3d | ASCII : %c\n", index, file[index], file[index], file[index], file[index], file[index])
+	// }
+
+	// BIG ENDIAN (starts with fe ff) (00 at the end is the null terminator)
+	// data is displayed in groups of 2 bytes
+	// each group representing a hexadecimal value
+	// byte order where most significant byte is stored first
+
+	// 0000000 [fe ff] [00 00 00 0a] 41 79 61 00 00 00 00 1e 50 72
+	// 0000020 69 73 68 61 00 ff ff ff ff 43 68 61 72 6c 69 65
+	// 0000040 00 00 00 00 19 4d 61 72 67 6f 74 [00]
+	// 0000054
+
+	// LITTLE ENDIAN (starts with ff fe) (00 at the end is the null terminator)
+	// data is displayed in groups of 2 bytes
+	// each group representing a hexadecimal value
+	// byte order where least significant byte is stored first
+
+	// 0000000 [ff fe] [0a 00 00 00] 41 79 61 00 1e 00 00 00 50 72
+	// 0000020 69 73 68 61 00 ff ff ff ff 43 68 61 72 6c 69 65
+	// 0000040 00 19 00 00 00 4d 61 72 67 6f 74 [00]
+	// 0000054
+
+	// initialise two flags (explicitly two, yes could be achieved with one)
+	var bigEndian bool
+	var littleEndian bool
+
+	// check the first two bytes of the file (deliberately using byte() to be explicit)
+	// if they are "fe ff" it is big endian
+	// if they are "ff fe" is is little endian
+	if file[0] == byte(0xfe) && file[1] == byte(0xff) {
+		bigEndian = true
+	} else if file[0] == byte(0xff) && file[1] == byte(0xfe) {
+		littleEndian = true
+	} else {
+		return errors.New("[3] could not establish Endianness of the binary file")
+	}
+
+	// set the start pointer to 2 to skip the first two bytes (which contain the Byte Order Mark BOM)
+	// then loop over the byte slice
+	for startIndex := 2; startIndex < len(file); {
+		// initialise a variable to store the high score
+		var highScore int32
+
+		// parse the high score (signed 32-bit integer) based on the Endianness
+		// (!) but why are we using Uint32 method (when we know its "signed") and then converting it to int32 (presumably "signed").. where is the native xEndian.Int32 method?
+		if bigEndian {
+			highScore = int32(binary.BigEndian.Uint32(file[startIndex : startIndex+4]))
+		} else if littleEndian {
+			highScore = int32(binary.LittleEndian.Uint32(file[startIndex : startIndex+4]))
+		}
```

Comment:
This feels quite repetitive - you're doing all of:
* `highSchore = ...`
* `int32(...)`
* `....Uint32(file[startIndex : startIndex+4])`

exactly the same in both branches. Can you think how you could avoid this duplication?

Reply:
I have found and used `binary.ByteOrder` which I didn't know before

highScore := int32(byteOrder.Uint32(file[startIndex : startIndex+4]))